src/extensions/default/NavigationAndHistory/NavigationProvider.js

```diff
@@ -0,0 +1,553 @@
+            } else {
+                this.selections.push({start: selection.from, end: selection.to});
+            }
+        }
+    };
+    
+   /**
+    * Function to clean up the markers in cm
+    */
+    NavigationFrame.prototype._clearMarkers = function () {
+        if (!this.cm) {
+            return;
+        }
+        var self = this;
+        
+        // clear only the markers we used to mark selections/cursors
+        this.cm.getAllMarks().filter(function (entry) {
+            if (entry.className === self.uId || self.bookMarkIds.indexOf(entry.id) !== -1) {
+                entry.clear();
+            }
+        });
+    };
+    
+    /**
+    * Function to check if we have valid markers in cm for this frame
+    */
+    NavigationFrame.prototype._validateMarkers = function () {
+        this._backupSelectionRanges();
+        return this.selections.length;
+    };
+
+   /**
+    * Function to actually navigate to the position(file,selections) captured in this frame
+    */
+    NavigationFrame.prototype.goTo = function () {
+        var self = this;
+        this._backupSelectionRanges();
+        jumpInProgress = true;
+        
+        // To ensure we don't reopen the same doc in the last known pane
+        // rather bring it to the same pane where user has opened it
+        var thisDoc = DocumentManager.getOpenDocumentForPath(this.file);
+        if (thisDoc && thisDoc._masterEditor) {
+            this.paneId = thisDoc._masterEditor._paneId;
+        }
+        
+        CommandManager.execute(Commands.FILE_OPEN, {fullPath: this.file, paneId: this.paneId}).done(function () {
+            EditorManager.getCurrentFullEditor().setSelections(self.selections, true);
+            _validateNavigationCmds();
+        }).always(function () {
+            jumpInProgress = false;
+        });
+    };
+    
+    
+   /**
+    * Function to capture a non-zero set of selections as a navigation frame.
+    * The assumptions behind capturing a frame as a navigation frame are :
+    *
+    * -> If it's set by user explicitly (using mouse click or jump to definition)
+    * -> By clicking on search results
+    * -> Change of cursor by keyboard navigation keys or actual edits are not captured.
+    *
+    * @private
+    */
+    function _recordJumpDef(event, selectionObj) {
+        // Don't capture frames if we are navigating or document text is being refreshed(fileSync in progress)
+        if (jumpInProgress || (event.target && event.target.document._refreshInProgress)) {
+            return;
+        }
+        // Reset forward navigation stack if we are capturing a new event
+        jumpedPosStack = [];
+        if (captureTimer) {
+            window.clearTimeout(captureTimer);
+            captureTimer = null;
+        }
+        
+        // Ensure cursor activity has not happened because of arrow keys or edit
+        if (selectionObj.origin !== "+move" && (!window.event || (window.event && window.event.type !== "input"))) {
```

Comment:
Refactored this code earlier but missed this condition completely. Changed it now.